### PR TITLE
Discover bots based on the entry points

### PR DESCRIPTION
**Pull request by kamil-certat**
The bot discovery algorithm was rewritten to use entry point
names instead of the file structure. This allow really easy
developing of packages with custom bots.

Note that the 'group' feature of entry points wasn't used
because we require bots to register callable scripts that
are later used to start bot's processes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,7 @@ CHANGELOG
 ### Core
 - `intelmq.lib.message`: For invalid message keys, add a hint on the failure to the exception: not allowed by configuration or not matching regular expression (PR#2398 by Sebastian Wagner).
 - `intelmq.lib.exceptions.InvalidKey`: Add optional parameter `additional_text` (PR#2398 by Sebastian Wagner).
+- Change the way we discover bots to allow easy extending based on the entry point name. (PR by Kamil Mankowski)
```
Comment (sebix) [inline CHANGELOG.md]:
```suggestion
- Change the way we discover bots to allow easy extending based on the entry point name. (PR#2413 by Kamil Mankowski)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +32,7 @@ CHANGELOG
 
 ### Documentation
 - Add a readthedocs configuration file to fix the build fail (PR#2403 by Sebastian Wagner).
+- Add a guide of developing extensions packages (PR by Kamil Mankowski)
```
Comment (sebix) [inline CHANGELOG.md]:
```suggestion
- Add a guide of developing extensions packages (PR#2413 by Kamil Mankowski)
```

File: contrib/example-extension-package/mybots/bots/collectors/custom/collector.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+"""
+SPDX-FileCopyrightText: 2023 CERT.at GmbH <https://cert.at/>
+SPDX-License-Identifier: AGPL-3.0-or-later
+"""
+
+from intelmq.lib.bot import CollectorBot
```
Comment (sebix) [inline contrib/example-extension-package/mybots/bots/collectors/custom/collector.py:9]: The example could additionally show, that import from a own library can be done like this:
```suggestion
from intelmq.lib.bot import CollectorBot
from mybots.lib import common_code
```
or alike

File: contrib/example-extension-package/mybots/bots/collectors/custom/collector.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class ExampleAdditionalCollectorBot(CollectorBot):
+    """
+    This is an example bot provided by extension package
```
Comment (sebix) [inline contrib/example-extension-package/mybots/bots/collectors/custom/collector.py]:
```suggestion
    This is an example bot provided by an extension package
```

File: contrib/example-extension-package/setup.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+"""
+
+from pathlib import Path
+from setuptools import find_packages, setup
+
+
+# Instead of the bot-autodiscovery below, you can also just manually declare entrypoints
+# (regardless of packaging solution, even in pyproject.toml etc.), e.g.:
+#
+# intelmq.bots.collectors.custom.collector = mybots.bots.collectors.custom.collector:BOT.run
```
Comment (sebix) [inline contrib/example-extension-package/setup.py]: it's a string
```suggestion
# 'intelmq.bots.collectors.custom.collector = mybots.bots.collectors.custom.collector:BOT.run'
```

File: docs/dev/guide.rst [hunk 6]
```diff
@@ -891,6 +891,66 @@ The databases `<` 10 are reserved for the IntelMQ core:
  * 3: statistics
  * 4: tests
 
+****************************
+Creating extensions packages
+****************************
+
+IntelMQ supports adding additional bots using your own independent packages. You can use this to
+add a new integration that is special to you, or cannot be integrated for some reason
+into the main IntelMQ repository.
```
Comment (sebix) [inline docs/dev/guide.rst]:
```suggestion
IntelMQ supports adding additional bots using your own independent packages. You can use this to
add a new integration that is special to you, or cannot be integrated
into the main IntelMQ repository for some reason.
```

File: docs/dev/guide.rst [hunk 7]
```diff
@@ -891,6 +891,66 @@ The databases `<` 10 are reserved for the IntelMQ core:
...
+
+Building an extension package
+=============================
+
+A simple example of the package can be found in ``contrib/example-extension-package``. In order to
+bots to work with IntelMQ, you need to ensure that
```
Comment (sebix) [inline docs/dev/guide.rst]:
```suggestion
A simple example of the package can be found in ``contrib/example-extension-package``. To make your custom
bots work with IntelMQ, you need to ensure that
```

File: docs/dev/guide.rst [hunk 8]
```diff
@@ -891,6 +891,66 @@ The databases `<` 10 are reserved for the IntelMQ core:
...
+
+ - your bot's module exposes a ``BOT`` object of the class inherited from ``intelmq.lib.bot.Bot``
+   or its subclasses,
+ - your package registers an `entry point <https://packaging.python.org/en/latest/specifications/entry-points/>`_
+   in the ``console_scripts`` group with a name starting with ``intelmq.bots.`` followed by
+   the name of the group (collectors, experts, outputs, parsers), and then your original name.
+   The entry point must point to the ``BOT.run`` method,
+ - the module in which the bot resides must be importable by IntelMQ (e.g. installed in the same
+   virtualv).
```
Comment (sebix) [inline docs/dev/guide.rst]:
```suggestion
   virtualenv, if you use them).
```

File: docs/dev/guide.rst [hunk 9]
```diff
@@ -891,6 +891,66 @@ The databases `<` 10 are reserved for the IntelMQ core:
...
+
+Apart from these requirements, your package can use any of the usual package features. We strongly
+recommend following the same principles and main guidelines as the official bots. This will ensure
+experience when using official and additional bots.
```
Comment (sebix) [inline docs/dev/guide.rst]: What kind of experience does it ensure?

File: docs/dev/guide.rst [hunk 10]
```diff
@@ -891,6 +891,66 @@ The databases `<` 10 are reserved for the IntelMQ core:
...
+
+The `pyproject.toml <https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#entry-points>`_
+file would then have the following section:
+
+.. code-block:: ini
+
+   [project.scripts]
+   intelmq.bots.collectors.awesome.special = "awesome_bots.special:BOT.run"
+
+Once you installed your package, you can run ``intelmqctl list bots`` to check if your bot was
```
Comment (sebix) [inline docs/dev/guide.rst]:
```suggestion
Once you have installed your package, you can run ``intelmqctl list bots`` to check if your bot was
```

File: intelmq/lib/utils.py [hunk 11]
```diff
@@ -860,13 +873,11 @@ def list_all_bots() -> dict:
     from intelmq.lib.bot import Bot  # noqa: prevents circular import
     bot_parameters = dir(Bot)
 
-    base_path = resource_filename('intelmq', 'bots')
-
-    botfiles = [botfile for botfile in pathlib.Path(base_path).glob('**/*.py') if botfile.is_file() and botfile.name != '__init__.py']
-    for file in botfiles:
-        file = Path(file.as_posix().replace(base_path, 'intelmq/bots'))
+    bot_entrypoints = filter(lambda entry: entry.name.startswith("intelmq.bots."), _get_console_entry_points())
```
Comment (sebix) [inline intelmq/lib/utils.py:889]: Is this covered by unit tests?

File: setup.py [hunk 12]
```diff
@@ -17,6 +17,7 @@
     'redis>=2.10',
     'requests>=2.2.0',
     'ruamel.yaml',
+    'importlib-metadata; python_version < "3.8"'
```
Comment (sebix) [inline setup.py:20]: Debian 10 buster has Python 3.7 and `python3-importlib-metadata` is only in buster-backports. So please update the installation instructions, adding that activating backports on Debian 10 is necessary and add the package as dependency to `debian/control`. I think it's totally fine to require backports on an old OS version.

**Review by sebix — CHANGES_REQUESTED** [review]

File: docs/dev/guide.rst (hunk 9 above)
Comment (kamil-certat) [reply docs/dev/guide.rst]: Looks like I lost some words when copying from language correction tool...

File: intelmq/lib/utils.py (hunk 11 above)
Comment (kamil-certat) [reply intelmq/lib/utils.py:889]: It was indirect tested (we test loading bots, and we have additional entry points in the package). But I've added a direct test for this case. Unfortunately, it's not easy to do it clear without mocking.

File: docs/user/installation.rst [hunk 13]
```diff
@@ -45,6 +45,10 @@ Native deb/rpm packages
 
 These are the operating systems which are currently supported by packages:
 
+* **Debian 10** Buster
+
+  * Enable the backport repository by the line ``deb-src http://deb.debian.org/debian buster-backports main contrib non-free`` to the file ``/etc/apt/sources.list`` first.
```
Comment (sebix) [inline docs/user/installation.rst]:
```suggestion
  * Enable the backport repository by the line ``deb http://deb.debian.org/debian buster-backports main contrib non-free`` to the file ``/etc/apt/sources.list`` first.
```

**Review by sebix — CHANGES_REQUESTED** [review]

File: docs/user/installation.rst (hunk 13 above)
Comment (sebix) [reply docs/user/installation.rst]: and is non-free required?

Comment (kamil-certat) [reply docs/user/installation.rst]: Thanks for mentioning this, I've failed yesterday on verifying in which part the package exactly ist. But today I've managed to find out, that only `main` is necessary

**Comment by kamil-certat** [comment]
After testing in our environment - fixes to places, where bot modules are imported

**Review by sebix — APPROVED** [review]

**Comment by sebix** [comment]
For further tests on more platforms I pushed your branch (with #2416) to the "unstable" package repo: https://build.opensuse.org/package/show/home:sebix:intelmq:unstable/intelmq

File: debian/control [hunk 14]
```diff
@@ -41,6 +41,7 @@ Depends: bash-completion,
          python3-ruamel.yaml,
          python3-termstyle (>= 0.1.10),
          python3-tz,
+         python3-importlib-metadata,
```
Comment (sebix) [inline debian/control:44]: In Debian 10 buster, this package is only available in backports. Operators need to enable this repository, for installing the IntelMQ deb package.

Comment (sebix) [reply debian/control:44]: we don't support Debian 10 anymore, so that is not relevant

**Comment by kamil-certat** [comment]
Thanks for the unstable package - I'll try to test the integration with the DEB package in the next two weeks. If you had your own use cases for tests, it would be great if you also could do some tests. From my side - the change is now used and works great for in our systems.

**Comment by kamil-certat** [comment]
@sebix I have just tested the integration with DEB package (IntelMQ from deb, additional bots from pip) - it works well! As it also works good in our instances for w few weeks already, I think we can merge it :)

**Pull request merged**